storage/memetcd/peer_store.go

```diff
@@ -0,0 +1,535 @@
+			}
+		}()
+	} else {
+		if err := ps.etcdDeletePeer(ih, p, true); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (ps *peerStore) PutLeecher(ih bittorrent.InfoHash, p bittorrent.Peer) error {
+	select {
+	case <-ps.closed:
+		panic("attempted to interact with stopped memory store")
+	default:
+	}
+
+	if err := ps.memoryStore.PutLeecher(ih, p); err != nil {
+		return err
+	}
+
+	if ps.cfg.EtcdOpAsync {
+		go func() {
+			if err := ps.etcdPutOrUpdatePeer(ih, p, false); err != nil {
+				log.Error(err)
+			}
+		}()
+	} else {
+		if err := ps.etcdPutOrUpdatePeer(ih, p, false); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (ps *peerStore) DeleteLeecher(ih bittorrent.InfoHash, p bittorrent.Peer) error {
+	select {
+	case <-ps.closed:
+		panic("attempted to interact with stopped memory store")
+	default:
+	}
+
+	if err := ps.memoryStore.DeleteLeecher(ih, p); err != nil {
+		return err
+	}
+
+	if ps.cfg.EtcdOpAsync {
+		go func() {
+			if err := ps.etcdDeletePeer(ih, p, false); err != nil {
+				log.Error(err)
+			}
+		}()
+	} else {
+		if err := ps.etcdDeletePeer(ih, p, false); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (ps *peerStore) GraduateLeecher(ih bittorrent.InfoHash, p bittorrent.Peer) error {
+	select {
+	case <-ps.closed:
+		panic("attempted to interact with stopped memory store")
+	default:
+	}
+
+	if err := ps.memoryStore.GraduateLeecher(ih, p); err != nil {
+		return err
+	}
+
+	if ps.cfg.EtcdOpAsync {
+		go func() {
+			if err := ps.etcdGraduateLeecher(ih, p); err != nil {
+				log.Error(err)
+			}
+		}()
+	} else {
+		if err := ps.etcdGraduateLeecher(ih, p); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (ps *peerStore) AnnouncePeers(ih bittorrent.InfoHash, seeder bool, numWant int, announcer bittorrent.Peer) (peers []bittorrent.Peer, err error) {
+	select {
+	case <-ps.closed:
+		panic("attempted to interact with stopped memory store")
+	default:
+	}
+
+	peers, err = ps.memoryStore.AnnouncePeers(ih, seeder, numWant, announcer)
+	return
+}
+
+func (ps *peerStore) ScrapeSwarm(ih bittorrent.InfoHash, addressFamily bittorrent.AddressFamily) (resp bittorrent.Scrape) {
+	select {
+	case <-ps.closed:
+		panic("attempted to interact with stopped memory store")
+	default:
+	}
+
+	resp = ps.memoryStore.ScrapeSwarm(ih, addressFamily)
+	return
+}
+
+// We do serializable reads of announce keys (compared to linearizable by default) for lower latency
+// Obvious con is that stale reads might occur which we are fine with in this scenario
+func (ps* peerStore) Bootstrap(...interface{}) error {
+	ctx, cancel := context.WithTimeout(context.Background(), ps.cfg.BootstrapTimeout)
+	defer cancel()
+
+	rangeResp, err := ps.etcd.Get(ctx, EtcdAnnouncePrefix, clientv3.WithPrefix(), clientv3.WithSerializable())
+	if err != nil {
+		return fmt.Errorf("failed to do etcd operation, bootstrap of all infohash-peer announces: %v", err)
+	}
+
+	announces := make([]interface{}, 0, rangeResp.Count)
```

Comment:
slice of some type cannot be passed to a function expecting slice of interface{} as it is:
https://ahmet.im/blog/golang-take-slices-of-any-type-as-input-parameter/
allocation for type should not be a concern since mem occupied by announce value itself will be much higher

Reply:
yeah, I was thinking about pointer of `Announce` struct, but as it's a one-time activity I guess current impl seems fine as well.